primitives/state-machine/src/overlayed_changes.rs

```diff
@@ -498,6 +508,42 @@ impl OverlayedChanges {
 		)
 	}
 
+	/// Get an iterator over all child trees that are stored in the overlay.
+	pub fn child_infos(&self) -> impl Iterator<Item=&ChildInfo> {
+		self.prospective.children_default.iter()
+			.chain(self.committed.children_default.iter())
+			.map(|(_, v)| &v.1)
+	}
+
+	/// Get an iterator over all pending and committed changes.
+	///
+	/// Supplying `None` for `child_info` will only return changes that are in the top
+	/// tree. Specifying some `child_info` will return only the changes in that
+	/// child tree.
+	pub fn changes(&self, child_info: Option<&ChildInfo>) -> impl Iterator<Item=(&StorageKey, &OverlayedValue)> {
+		let (committed, prospective) = if let Some(child_info) = child_info {
+			match child_info.child_type() {
+				ChildType::ParentKeyId => (
+					self.committed.children_default.get(child_info.storage_key()).map(|c| &c.0),
+					self.prospective.children_default.get(child_info.storage_key()).map(|c| &c.0),
+				),
+			}
+		} else {
+			(Some(&self.committed.top), Some(&self.prospective.top))
+		};
```

Comment:
nit: the single `Option` parameter which changes the behavior of this function significantly suggest that maybe we should separate this function on two? A drawback that I see are minor code duplication for the line below which IMO. But doc and logic would be a bit simpler and the call sites would be a bit tider.

Reply:
As a matter of fact I chose this approach because it simplifies the calling site where the code is the same for child trees and top trees. Splitting that up would require the caller to chain top and child trie changes together.

That said, I actually think splitting would be the right thing to do I will investigate.

Reply:
Ah, I didn't account for that fact. Up to you then!

Reply:
Put some more thought into it. I cannot really come up with something better that satisfies every user. I will leave it as is in absence of further arguments.